File: src/main/java/net/obesephoenix/koalakraftcrystals/events/onItemDespawn.java
package net.obesephoenix.koalakraftcrystals.events;

import net.obesephoenix.koalakraftcrystals.KoalaKraftCrystals;
import org.bukkit.NamespacedKey;
import org.bukkit.entity.Item;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.ItemDespawnEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.persistence.PersistentDataType;

public class onItemDespawn implements Listener {

    @EventHandler
    public void onEvent(ItemDespawnEvent event) {
        Item entity = event.getEntity();
        ItemStack item = entity.getItemStack();
        ItemMeta meta = item.getItemMeta();
        if (meta == null) return;

        if (meta.getPersistentDataContainer().has(new NamespacedKey(KoalaKraftCrystals.instance, "crystal_id")
                , PersistentDataType.STRING)) {
            event.setCancelled(true);
        }
    }
}

File: src/main/kotlin/net/obesephoenix/koalakraftcrystals/events/OnItemDespawn.kt
package net.obesephoenix.koalakraftcrystals.events

import org.bukkit.event.entity.ItemDespawnEvent
import org.bukkit.inventory.meta.ItemMeta
import org.bukkit.NamespacedKey
import net.obesephoenix.koalakraftcrystals.KoalaKraftCrystals
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.persistence.PersistentDataType

class OnItemDespawn : Listener {
    @EventHandler
    fun onEvent(event: ItemDespawnEvent) {
        val entity = event.entity
        val item = entity.itemStack
        val meta = item.itemMeta ?: return
        if (meta.persistentDataContainer.has(
                NamespacedKey(KoalaKraftCrystals.instance, "crystal_id"), PersistentDataType.STRING
            )
        ) {
            event.isCancelled = true
        }
    }
}